File: example/LobbyServer/Component/Dungeon/Unit.cpp
#include "Unit.h"
#include <atomic>
#include "Dungeon.h"
#include <idl/MessageDungeon.h>
#include "../../UserSession.h"
#include "Unit/Player.h"
#include "Unit/Monster.h"

namespace Component { namespace Unit {
static std::atomic<uint64_t> UNIT_SEQ;

Meta::Meta()
	: id("")
	, index(0)
	, max_health(0)
	, attack(0)
	, defense(0)
	, speed(0)
	, sight(0)
	, critical_chance(0.0f)
	, critical_rate(0.0f)
{
	META_MEMBER(id);
	META_MEMBER(index);
	META_MEMBER(monster);
}

Data::Data()
	: dungeon(nullptr)
	, seq(++UNIT_SEQ)
	, max_health(0)
	, cur_health(0)
	, attack(0)
	, defense(0)
	, speed(0)
	, sight(10)
	, critical_chance(0.0f)
	, critical_rate(0.0f)
	, timer(Gamnet::Time::RepeatTimer::Create())
{
}

void Data::SetPosition(const Vector2Int& position)
{
	std::shared_ptr<Tile> nextTile = dungeon->GetTile(position);
	if (nullptr == nextTile)
	{
		return;
	}

	if(nullptr != nextTile->unit && shared_from_this() != nextTile->unit)
	{
		return;
	}

	std::shared_ptr<Tile> curTile = dungeon->GetTile(this->position);
	if(nullptr != curTile)
	{
		curTile->unit = nullptr;
	}

	this->position = position;
	nextTile->unit = shared_from_this();
}

void Manager::Init()
{
	InitMeta("../MetaData/Unit_Player.csv");
	InitMeta("../MetaData/Unit_Monster.csv");
}

void Manager::InitMeta(const std::string& path)
{
	MetaReader<Meta> reader;
	auto& rows = reader.Read(path);
	for (auto& row : rows)
	{
		if (false == id_metas.insert(std::make_pair(row->id, row)).second)
		{
			throw GAMNET_EXCEPTION(Message::ErrorCode::UndefineError, "duplicate item id(", row->id, ")");
		}
		if (false == index_metas.insert(std::make_pair(row->index, row)).second)
		{
			throw GAMNET_EXCEPTION(Message::ErrorCode::UndefineError, "duplicate item index(", row->index, ")");
		}
	}
}
std::shared_ptr<Meta> Manager::FindMeta(const std::string& id)
{
	auto itr = id_metas.find(id);
	if (id_metas.end() == itr)
	{
		return nullptr;
	}
	return itr->second;
}

std::shared_ptr<Meta> Manager::FindMeta(uint32_t index)
{
	auto itr = index_metas.find(index);
	if (index_metas.end() == itr)
	{
		return nullptr;
	}
	return itr->second;
}

std::shared_ptr<Unit::Data> Manager::CreateInstance(const std::string& id)
{
	std::shared_ptr<Meta> meta = FindMeta(id);
	if (nullptr == meta)
	{
		return nullptr;
	}

	return CreateInstance(meta);
}

std::shared_ptr<Unit::Data> Manager::CreateInstance(uint32_t index)
{
	std::shared_ptr<Meta> meta = FindMeta(index);
	if (nullptr == meta)
	{
		return nullptr;
	}

	return CreateInstance(meta);
}

std::shared_ptr<Unit::Data> Manager::CreateInstance(const std::shared_ptr<Meta>& meta)
{
	std::shared_ptr<Unit::Data> unit = std::make_shared<Unit::Data>();
	unit->meta = meta;
	if(nullptr != unit->meta->monster)
	{
		unit->AddComponent<Monster::Data>(std::make_shared<Monster::Data>(unit->meta->monster));
	}
	return unit;
}

std::shared_ptr<Unit::Data> CreatePlayer(std::shared_ptr<UserSession> session, const std::string& id)
{
	if(nullptr != session->GetComponent<Unit::Data>())
	{
		throw GAMNET_EXCEPTION(Message::ErrorCode::UndefineError);
	}
	std::shared_ptr<Unit::Data> data = Gamnet::Singleton<Manager>::GetInstance().CreateInstance(id);
	data->team_index = 0;
	data->AddComponent<Player>();
	session->AddComponent<Unit::Data>(data);
	return data;
}
}}

GAMNET_BIND_INIT_HANDLER(Component::Unit::Manager, Init);

File: example/unity_client/Assets/Scripts/Scene/Dungeon/Unit.cs
using UnityEngine;
using System.Collections.Generic;
using System;

public class Unit : MonoBehaviour
{
	public Component.Dungeon dungeon;
	public UInt64 seq;
	public int max_health;
	public int cur_health;
	public int attack;
	public int defense;
	public int speed;
	public int sight;
	public float critical_chance;
	public float critical_rate;

	public Vector2Int position
	{
		get
		{
			return new Vector2Int((int)transform.position.x, (int)transform.position.y);
		}
		set
		{
			if (true == GetComponent<PlayerController>().enabled)
			{
				GameManager.Instance.dungeon.SetFieldOfView(position, sight, false);
				GameManager.Instance.dungeon.SetFieldOfView(value, sight, true);
			}
			else
			{
				Tile tile = null;
				tile = GameManager.Instance.dungeon.GetTileData(position.x, position.y);
				if (null != tile)
				{
					tile.Leave(this);
				}
				tile = GameManager.Instance.dungeon.GetTileData(value.x, value.y);
				if (null != tile)
				{
					tile.Enter(this);
				}
			}
			transform.position = new Vector3(value.x, value.y, 0.0f);
		}
	}

	public System.Action<float> onDamage;

	public void SetMovePath(List<Vector2Int> path)
	{
		this.path = path;
		this.delta_time = 1.0f;
	}

	private List<Vector2Int> path = null;
	private float delta_time = 0.0f;

	private void Update()
	{
		if (null != path)
		{
			if (0 >= path.Count)
			{
				path = null;
				return;
			}

			delta_time += Time.deltaTime;
			if (GameManager.TickTime > delta_time)
			{
				return;
			}

			delta_time = 0.0f;
			position = path[0];
			path.RemoveAt(0);
		}
	}

	public class Meta : MetaData
	{
		public string id;
		public UInt32 index;

		public Meta()
		{
		}
	}

	public class Manager : Util.Singleton<Manager>
	{
		public Manager()
		{
			//GameManager.Instance.LobbySession.RegisterHandler<Message.Dungeon.MsgSvrCli_UnitCreate_Ntf>(OnUnitCreate);
			//GameManager.Instance.LobbySession.RegisterHandler<Message.Dungeon.MsgSvrCli_UnitPosition_Ntf>(OnUnitPosition);
		}

		public void Init()
		{
			InitMeta("Assets/MetaData/Unit_Monster.csv");
		}
		private void InitMeta(string filePath)
		{
			TextAsset textAsset = AssetBundleManager.Instance.LoadAsset<TextAsset>(filePath);
			if (null == textAsset)
			{
				GameManager.Instance.ui.alert.Open("Failed to load meta file", "can not find '" + filePath + "'");
				throw new System.Exception();
			}
			CSVReader csvReader = new CSVReader();
			csvReader.ReadStream(textAsset.text);
			MetaData.Reader<Meta> metaReader = new MetaData.Reader<Meta>();
			metaReader.Read(csvReader);
			foreach (Meta meta in metaReader)
			{
				if (true == id_metas.ContainsKey(meta.id))
				{
					throw new System.Exception("duplicate item id(" + meta.id + ")");
				}

				if (true == index_metas.ContainsKey(meta.index))
				{
					throw new System.Exception("duplicate item index(" + meta.index + ")");
				}

				id_metas.Add(meta.id, meta);
				index_metas.Add(meta.index, meta);
			}
		}

		private Dictionary<UInt32, Meta> index_metas = new Dictionary<UInt32, Meta>();
		private Dictionary<string, Meta> id_metas = new Dictionary<string, Meta>();
	}
}
